src/types.jl

```diff
@@ -116,123 +116,56 @@ function show(io::IO, state::Step)
-                     # TODO: we need a better guess here, possibly
-                     # overwrite it in the call to solve()
-                     initstep = minstep,
-                     norm     = Base.norm,
-                     maxiters = T(Inf),
-                     points   = :all,
-                     stopevent = (t,y)->false,
-                     roottol  = eps(T)^T(1//3),
-                     isoutofdomain = isnan,
-                     kargs...)
-        if all(points .!= [:specified,:all])
-            error("Option points = $points is not supported, use :specified or :all")
-        end
-        #TODO iterate over fields here?
-        new(initstep,tstop,reltol,abstol,minstep,maxstep,norm,maxiters,isoutofdomain,sort(tspan),points,stopevent,roottol)
-    end
-
-end
-
-function show{T}(io::IO, opts :: Options{T})
-    for name in fieldnames(opts)
-        @printf("%-20s = %s\n",name,getfield(opts,name))
-    end
-end
-
+abstract AbstractSolver{T,Y}
 
 """
 
 This is an iterable type, each call to next(...) produces a next step
-of a numerical solution to an ODE.
+of a numerical solution to an ODE.  Types `T` and `Y` determine the
+output type (`Tuple{T,Y}`) for iteration.
 
 - ode: is the prescrived ode, along with the initial data
 - stepper: the algorithm used to produce subsequent steps
-- options: options passed to the stepper
 
 """
-immutable Solver{O<:AbstractIVP,S<:AbstractStepper,T}
+immutable Solver{O<:AbstractIVP,S<:AbstractStepper,T,Y} <: AbstractSolver{T,Y}
     ode     :: O
     stepper :: S
-    options :: Options{T}
 end
 
-# filter the wrong combinations of ode and stepper
-solve{T,S}(ode::T, stepper::S, options) = error("The $S doesn't support $T")
+Base.eltype{T,Y}(::Type{AbstractSolver{T,Y}}) = Tuple{T,Y}
 
+Solver{T,Y}(ode::AbstractIVP{T,Y},stepper::AbstractStepper{T}) =
+    Solver{typeof(ode),typeof(stepper),T,Y}(ode,stepper)
 
-# normally we return the working array, which changes at each step and
-# expect the user to copy it if necessary.  In order for collect to
-# return the expected result we need to copy the output at each step.
-function collect{T,S}(t::Type{Tuple{T,S}}, s::Solver)
-    if maximum(s.options.tspan) == T(Inf)
-        error("Attempting to collect an infinite list, use tstop or tspan with finite numbers only")
+# filter the wrong combinations of ode and stepper
+solve{T,S}(ode::T, stepper::Type{S}, options...) =
+    error("The $S doesn't support $T")
+
+# In Julia 0.5 the collect needs length to be defined, we cannot do
+# that for a solver
+function collect{T,Y}(s::AbstractSolver{T,Y})
+    pairs = Array(Tuple{T,Y},0)
+    for (t,y) in s
+        push!(pairs,(t,copy(y)))
     end
-    collect(t, imap(x->deepcopy(x),s))
+    return pairs
 end
 
-function collect(s::Solver)
-    if maximum(s.options.tspan) == Inf
-        error("Attempting to collect an infinite list, use tstop or tspan with finite numbers only")
+# Transpose a solver to get a (Vector(t),Vector(yout)) style output
```

Comment:
This could also be implemented as `collect_transposed` removing the need for an additional type.